### Q.3. Use the dataset Data1.csv to implement the list of classification models below in order to predict whether a tumor is malign or beginning. The dependent variable in this dataset is Class, it is equal to 2 if the tumor is benign and 4 if it is malign. The remaining columns are features that describe different characteristics of each tumor.

### Import the libraries

In [213]:
from sklearn import svm
import pandas as pd
from sklearn.model_selection import train_test_split
import numpy as np
import matplotlib.pyplot as plt
from sklearn import metrics
import seaborn as sb

### Read the data from the csv.

In [214]:
dataset = pd.read_csv('Data1.csv')
x = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values

### Now will split the dataset into the training set and test set

In [215]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 0)

### Feature scaling

In [216]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

### 1. Logistic Regression

Import the Logistic Regression library

In [217]:
from sklearn.linear_model import LogisticRegression

Create the Logistic Regression object.

In [218]:
logisticRegr = LogisticRegression(random_state=0)

Training the model on the data, storing the information learned from the data
Model is learning the relationship between digits (x_train) and labels (y_train).

In [219]:
logisticRegr.fit(x_train, y_train)

LogisticRegression(random_state=0)

Predict labels for new data. Uses the information the model learned during the model training process.

In [220]:
logisticRegr.predict(x_test[0].reshape(1,-1))

array([2], dtype=int64)

Predict for Multiple Observations at Once

In [221]:
logisticRegr.predict(x_test[0:10])

array([2, 2, 4, 4, 2, 2, 2, 4, 2, 2], dtype=int64)

Make predictions on entire test data.

In [222]:
predictions = logisticRegr.predict(x_test)

Let's see the accuracy of the model using the score method:

In [223]:
score = logisticRegr.score(x_test, y_test)
print("Accuracy: {:.2f} %".format(score*100))

Accuracy: 95.62 %


In [224]:
cm = metrics.confusion_matrix(y_test, predictions)
print(cm)

[[84  3]
 [ 3 47]]


According to the results above the accuracy of the model is not quite good - 95.62%. There were 6 values out of 137 that were not predicted correctly.

Next, using K-fold cross-validation method (cv = 10) will ensure that the accuracy score calculated only on one test set was not just a lucky occurence. Will print the nest accuracy score and the standard deviation  of the score computed. 

In [225]:
from sklearn.model_selection import cross_val_score
logisticRegr_accuracies = cross_val_score(estimator = logisticRegr, X = x_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(logisticRegr_accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(logisticRegr_accuracies.std()*100))
logisticRegr_accuracies

Accuracy: 96.70 %
Standard Deviation: 1.97 %


array([0.94545455, 0.96363636, 0.96363636, 1.        , 0.94545455,
       1.        , 0.96296296, 0.96296296, 0.98148148, 0.94444444])

The mean accurracy for 10 test sets is 96.7% that is not much bigger that the accuracy of the model calculated on only one train data set (95.62%).

### 2. Decision Tree Classifier

Let's create a Decision Tree Model using Scikit-learn. Import Decision Tree Classifier

In [226]:
from sklearn.tree import DecisionTreeClassifier

Create Decision Tree classifer object.

In [227]:
dtc = DecisionTreeClassifier( random_state=0)

Train Decision Tree Classifer.

In [228]:
dtc = dtc.fit(x_train, y_train)

Predict the response for test dataset.

In [229]:
y_pred = dtc.predict(x_test)

Let's estimate, how accurately the classifier or model can predict whether a tumor is malign or bening. Accuracy can be computed by comparing actual test set values and predicted values.



In [230]:
print("Accuracy: {:.2f} %".format(metrics.accuracy_score(y_test, y_pred)*100))

Accuracy: 92.70 %


In [231]:
cm = metrics.confusion_matrix(y_test, y_pred)
print(cm)

[[80  7]
 [ 3 47]]


We see that this model has a big accuracy that means it is good to use for predicting whether a tumor is malign or benign. The confusion matrix also shows that only 10 out of 137 objects are not predicted as they had to.

Next, using K-fold cross-validation method (cv = 10) will ensure that the accuracy score calculated only on one test set was not just a lucky occurence. Will print the nest accuracy score and the standard deviation  of the score computed. 

In [232]:
from sklearn.model_selection import cross_val_score
dtc_accuracies = cross_val_score(estimator = dtc, X = x_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(dtc_accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(dtc_accuracies.std()*100))
dtc_accuracies

Accuracy: 93.59 %
Standard Deviation: 2.21 %


array([0.92727273, 0.90909091, 0.92727273, 0.96363636, 0.94545455,
       0.92727273, 0.90740741, 0.92592593, 0.98148148, 0.94444444])

The mean accuracy calculated for 10 subsets of data is bigger with 0.89% that the accuracy of the model calculated on only one test set.

### 3. Random Forest Classifier (with nb_trees = 10)

Import Random Forest Model.

In [233]:
from sklearn.ensemble import RandomForestClassifier

Create a Gaussian Classifier.

In [234]:
rfc=RandomForestClassifier(n_estimators=10)


Train the model using the training sets y_pred=clf.predict(X_test)


In [235]:
rfc.fit(x_train,y_train)

RandomForestClassifier(n_estimators=10)

Predict the response for test dataset.

In [236]:
y_pred=rfc.predict(x_test)

After training, check the accuracy using actual and predicted values.

In [237]:
print("Accuracy: {:.2f} %".format(metrics.accuracy_score(y_test, y_pred)*100))

Accuracy: 93.43 %


In [238]:
cm = metrics.confusion_matrix(y_test, y_pred)
print(cm)

[[83  4]
 [ 5 45]]


We can observe that the accuracy of the model is pretty high - 96,35%. There are only 5 values that were not predicted to the correct group.

Next, using K-fold cross-validation method (cv = 10) will ensure that the accuracy score calculated only on one test set was not just a lucky occurence. Will print the nest accuracy score and the standard deviation  of the score computed. 

In [239]:
from sklearn.model_selection import cross_val_score
rfc_accuracies = cross_val_score(estimator = rfc, X = x_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(rfc_accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(rfc_accuracies.std()*100))
rfc_accuracies

Accuracy: 97.08 %
Standard Deviation: 2.02 %


array([0.94545455, 0.94545455, 0.98181818, 0.96363636, 0.94545455,
       1.        , 0.98148148, 0.96296296, 1.        , 0.98148148])

Here we can observe that the mean value of the accuracy for the model calculated on 10 test datasets (96.15%) is bigger than the value of the accuracy on the model calculated on only one test dataset (95.62%). 

### 4. K- Nearest Neighbors (K-NN)

Import the KNeighborsClassifier library

In [240]:
from sklearn.neighbors import KNeighborsClassifier

Create the object of KNeighborsClassifier type with number of neighbors =3.

In [241]:
neigh = KNeighborsClassifier(n_neighbors=3)

Train the KNeighbors Classifier.

In [242]:
neigh.fit(x_train,y_train)

KNeighborsClassifier(n_neighbors=3)

Predict the response for test dataset.

In [243]:
y_pred = neigh.predict(x_test)

After training, check the accuracy using actual and predicted values.

In [244]:
print("Accuracy: {:.2f} %".format(metrics.accuracy_score(y_test, y_pred)*100))

Accuracy: 97.08 %


In [245]:
cm = metrics.confusion_matrix(y_test, y_pred)
print(cm)

[[85  2]
 [ 2 48]]


We can observe that this model is better that the previous ones. This has 97.08% of accuracy and only 4 values out of 137 are predicted wrongly. For sure is one of the best candidates for the best models.

Next, using K-fold cross-validation method (cv = 10) will ensure that the accuracy score calculated only on one test set was not just a lucky occurence. Will print the nest accuracy score and the standard deviation  of the score computed. 

In [246]:
from sklearn.model_selection import cross_val_score
neigh_accuracies = cross_val_score(estimator = neigh, X = x_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(neigh_accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(neigh_accuracies.std()*100))
neigh_accuracies

Accuracy: 96.34 %
Standard Deviation: 2.44 %


array([0.90909091, 0.94545455, 0.98181818, 0.98181818, 0.94545455,
       1.        , 0.96296296, 0.96296296, 0.98148148, 0.96296296])

We can observe that the accuracy for the model calculated on only one test dataset is bigger than the mean accuracy for the model calculated on 10 smaller test data sets with 0.74%. Here we can say that there was a little lucky occurences.

### 5. Naive Bayes

Import the GaussianNB library

In [247]:
from sklearn.naive_bayes import GaussianNB

Create the GaussianNB object and train it on the train data.

In [248]:
gnb = GaussianNB()
gnb.fit(x_train, y_train)

GaussianNB()

Making predictions on the testing set


In [249]:
y_pred = gnb.predict(x_test)

Comparing actual response values (y_test) with predicted response values (y_pred)


In [250]:
from sklearn import metrics
print("Accuracy: {:.2f}%".format(metrics.accuracy_score(y_test, y_pred)*100))

Accuracy: 94.89%


In [251]:
cm = metrics.confusion_matrix(y_test, y_pred)
print(cm)

[[80  7]
 [ 0 50]]


The model has a pretty good accuracy- 94.89%. There are only 7 values that were predicted wrongly.

Using K-fold cross-validation method (cv = 10) will ensure that the accuracy score calculated only on one test set was not just a lucky occurence. Will print the nest accuracy score and the standard deviation  of the score computed. 

In [252]:
from sklearn.model_selection import cross_val_score
gnb_accuracies = cross_val_score(estimator = gnb, X = x_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(gnb_accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(gnb_accuracies.std()*100))
gnb_accuracies

Accuracy: 96.16 %
Standard Deviation: 1.91 %


array([0.92727273, 0.96363636, 0.96363636, 0.96363636, 0.96363636,
       0.96363636, 0.98148148, 0.94444444, 1.        , 0.94444444])

There can be observed that the accuracy score of the model calculated on only one test set is smaller that the accuracy mean calculated on 10 smaller data sets. (94.89% < 96.16%)

### 6. Support Vector Machine (SVM)

Train the SVC classifier

In [253]:
clf = svm.SVC(random_state=0)
clf.fit(x_train, y_train)

SVC(random_state=0)

Make predictions and compare the results by displaying the predicted values of y next to the test values of y in a two-dimensional array.

In [254]:
y_pred = clf.predict(x_test)
y_pred

array([2, 2, 4, 4, 2, 2, 2, 4, 2, 2, 4, 2, 4, 2, 2, 4, 4, 4, 4, 2, 2, 2,
       4, 2, 4, 4, 2, 2, 2, 4, 2, 4, 4, 2, 2, 2, 4, 4, 2, 4, 2, 2, 2, 2,
       2, 2, 2, 4, 2, 2, 4, 2, 4, 2, 2, 2, 4, 4, 2, 4, 2, 2, 2, 2, 2, 2,
       2, 2, 4, 4, 2, 2, 2, 2, 2, 2, 4, 2, 2, 2, 4, 2, 4, 2, 2, 4, 2, 4,
       4, 2, 4, 2, 4, 4, 2, 4, 4, 4, 4, 2, 2, 2, 4, 4, 2, 2, 4, 4, 2, 2,
       4, 2, 2, 4, 2, 2, 2, 4, 2, 2, 2, 4, 2, 2, 4, 4, 2, 4, 2, 4, 4, 2,
       4, 2, 2, 4, 2], dtype=int64)

Create and print the Confusion Matrix and the accuracy score of the model and interpret the two.

In [255]:
from sklearn.metrics import confusion_matrix,accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
print("SVC model accuracy: {:.2f}%".format(accuracy_score(y_test, y_pred)*100))

[[82  5]
 [ 1 49]]
SVC model accuracy: 95.62%


We can observe that this is a good model with an accuracy of 95.62%. There are 6 values that were not predicted correctly.

Using K-fold cross-validation method (cv = 10) will ensure that the accuracy score calculated only on one test set was not just a lucky occurence. Will print the nest accuracy score and the standard deviation  of the score computed. 

In [256]:
from sklearn.model_selection import cross_val_score
svm_accuracies = cross_val_score(estimator = clf, X = x_train, y = y_train, cv = 10)
print("Accuracy: {:.2f} %".format(svm_accuracies.mean()*100))
print("Standard Deviation: {:.2f} %".format(svm_accuracies.std()*100))
svm_accuracies

Accuracy: 96.53 %
Standard Deviation: 1.91 %


array([0.92727273, 0.96363636, 0.96363636, 0.98181818, 0.96363636,
       0.96363636, 0.98148148, 0.96296296, 1.        , 0.94444444])

We can observe that the accuracy score calculated for a only one test set is smaller with 0.92% than the accuracy mean calculatd for 10 test sets.

## Q.4. Choose the best performing model based on the results from performing the k-fold cross-validation. Discuss your choice.

In [261]:
print("Mean accuracy for Logistic Regression model: {:.2f} %, Std: {:.2f}.".format(logisticRegr_accuracies.mean()*100, logisticRegr_accuracies.std()*100))
print("Mean accuracy for Decision Tree Classifier model: {:.2f}%, Std: {:.2f}.".format(dtc_accuracies.mean()*100, dtc_accuracies.std()*100))
print("Mean accuracy for Random Forest Classifier with 10 nb_trees: {:.2f} %, Std: {:.2f}.".format(rfc_accuracies.mean()*100, rfc_accuracies.std()*100))
print("Mean accuracy for K-Nearest Neighbors: {:.2f} %, Std: {:.2f}.".format(neigh_accuracies.mean()*100, neigh_accuracies.std()*100))
print("Mean accuracy for Naive Bayes: {:.2f} %, Std: {:.2f}.".format(gnb_accuracies.mean()*100, gnb_accuracies.std()*100))
print("Mean accuracy for Support Vector Machine: {:.2f} %, Std: {:.2f}.".format(svm_accuracies.mean()*100, svm_accuracies.std()*100))



Mean accuracy for Logistic Regression model: 96.70 %, Std: 1.97.
Mean accuracy for Decision Tree Classifier model: 93.59%, Std: 2.21.
Mean accuracy for Random Forest Classifier with 10 nb_trees: 97.08 %, Std: 2.02.
Mean accuracy for K-Nearest Neighbors: 96.34 %, Std: 2.44.
Mean accuracy for Naive Bayes: 96.16 %, Std: 1.91.
Mean accuracy for Support Vector Machine: 96.53 %, Std: 1.91.


We can observe that the best mean accuracy is for Random Forest Classifier model.  